0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,218 @@
+  0. The sender appends the private headers, followed by `\n\n`.
+  0. The sender appends the application data (in its raw binary format).
+0. The sender quotes and sends their payment, setting the data field of the ILP packet to the PSK data.
+0. The receiver gets notified by their ledger of an incoming prepared transfer with an ILP packet.
+0. The receiver generates the payment encryption key from the nonce contained in the PSK datas' public headers, and uses it to decrypt the private headers and their data.
+0. The receiver verifies that the amount in the incoming transfer matches the amount in the ILP packet. The receiver may also call a backend to make sure the incoming funds are expected.
+0. The receiver fulfills the incoming transfer with the HMAC of the ILP packet, using an HMAC key derived from the shared secret.
+
+## Data Format
+
+The PSK details are divided into two sets of headers, like those found in [HTTP Requests](https://tools.ietf.org/html/rfc7230#section-3.2). Application data is appended to the private headers, after a blank line. This object is then encrypted and appended to the public headers as binary data, after a blank line. Both priavte and public headers are parsed in the exact same way as HTTP headers.
+
+The public headers come after a single status line, which must be `PSK/1.0`. The public headers are visible to all parties transmitting the packet, so they SHOULD contain only the bare minimum required for the intended parties to decrypt and validate the private headers.
+
+### Public Headers
+
+The public headers and their data are formatted as below:
+
+```http
+PSK/1.0
+Key: hmac-sha-256 fpwpAhlN588
+Header: data_everyone_can_see
+
+...
+```
+
+Everything stored in these headers is readable by all connectors between the sender and receiver. PSK payments use an HMAC of the PSK details as a fulfillment, in order to prove that the contents are unmodified. The pre-shared secret key is the key used in the HMAC.
+
+The data attached to these headers is an encrypted binary blob. This encrypted blob contains the private PSK headers.
+
+Despite the fact that connectors can read the public headers' data, connectors should not be required to do so, nor should they rely on PSK details being present on all transfers. This information is only intended for the initial sender and final receiver of a payment.
+
+The only headers in the public headers should be the ones required for the receiver to decrypt the private headers. The decryption key can be calculated by providing the data and the key to the HMAC-SHA-256 function. In the above example, the data is `fpwpAhlN588`. The key is the pre-shared secret key.
+
+| Header | Value |
+|--------|-------|
+| `Key` | _(Required)_ The HMAC algorithm, which MUST be `hmac-sha-256` (case sensitive), followed by a base64url-encoded nonce, used to generate the decryption key for the private headers. |
```

Comment:
Link to a definition of base64url encoding, like https://en.wikipedia.org/wiki/Base64#URL_applications